### ISSUE TITLE
[feat] `additionalRefHooks` setting

### ISSUE THREAD

**Issue by silverwind** [Rel1cx/eslint-react#1904]
### Problem Description

It would be useful to expose a global plugin setting where users can specify ref-returning hooks, `react-x/immutability` and maybe other rules could benefit. My example for this is a `useRef` variant which accepts a initializer function, based on [this](https://thoughtspile.github.io/2021/11/30/lazy-useref/):

```tsx
const emptyValue = Symbol("useLazyRef emptyValue");

function useLazyRef<T>(init: () => T): RefObject<T> {
  const ref = useRef(emptyValue as T);
  if (ref.current === emptyValue) {
    ref.current = init();
  }
  return ref;
}
```

### Feature Details

A `additionalRefHooks` setting, similar to existing `additionalStateHooks` and `additionalEffectHooks`. Takes an array of ref-returning function names.

### Examples

It would make `react-x/immutability` no longer false-positive on this:

```tsx
export function LazyRefDemo() {
  const seen = useLazyRef(() => new Set<string>());
  useEffect(() => { seen.current.add("x"); }, [seen]);
  return null;
}
```

### Evaluation Checklist

- [x] I have had problems that this feature would solve
- [x] I could not find a way to solve the problem with existing features or workarounds
- [x] I have thought very hard about potential edge cases and downsides, and they are acceptable
- [x] I think the feature is well-defined and would provide clear value to users

**Comment by christopher-buss** [comment]
Another case for this, `immutability` also hardcodes a second exemption: `useHistory()`/`useNavigate()`/`useNavigation()`. These method calls aren't flagged because they're external stores rather than refs. That list has the same problem `additionalRefHooks` solves, but for a different category.

```tsx
export function useWorld(): World {
  return useContext(WorldContext)!;
}

function Component() {
  const world = useWorld(); // returns a external mutable store

  const onClose = useCallback(() => {
    world.set( { ... });
  }, [world]);

  return <AnotherComponent onClose={onClose} />;
}
```

Reported twice, correctly per the current contract: `world` is captured from component scope, `set` mutates, and the handler escapes render. But `world` is an external store exactly like a `navigate` object, where writing to it is intended.

`additionalRefHooks` would silence it if we registered `useWorld` there, but the world isn't a ref, so that just hides a misdeclaration in config. I'd propose a `additionalMutableHooks` covering the navigation-hook list would fit, I am happy to open a separate issue for this if it's better tracked separately than here.

**Comment by Rel1cx** [comment]
Sorry for the late reply. This feature request has been accepted, and I plan to introduce `additionalRefHooks` in the next minor release.

**Comment by Rel1cx** [comment]
> Another case for this, `immutability` also hardcodes a second exemption: `useHistory()`/`useNavigate()`/`useNavigation()`. These method calls aren't flagged because they're external stores rather than refs. That list has the same problem `additionalRefHooks` solves, but for a different category.
> 
> export function useWorld(): World {
>   return useContext(WorldContext)!;
> }
> 
> function Component() {
>   const world = useWorld(); // returns a external mutable store
> 
>   const onClose = useCallback(() => {
>     world.set( { ... });
>   }, [world]);
> 
>   return <AnotherComponent onClose={onClose} />;
> }
> Reported twice, correctly per the current contract: `world` is captured from component scope, `set` mutates, and the handler escapes render. But `world` is an external store exactly like a `navigate` object, where writing to it is intended.
> 
> `additionalRefHooks` would silence it if we registered `useWorld` there, but the world isn't a ref, so that just hides a misdeclaration in config. I'd propose a `additionalMutableHooks` covering the navigation-hook list would fit, I am happy to open a separate issue for this if it's better tracked separately than here.

Once the next minor release ships, you can use the `additionalRefHooks` setting as a temporary workaround. Later, once we land support similar to #1916 (a React Compiler `environment` config with per-hook schemas for custom hooks, though the concrete plan is still undecided), rules will be able to read the effect kind and value kind of custom hooks like `useWorld` from semantically correct dedicated properties, including whether their return values are frozen or mutable, instead of overloading `additionalRefHooks`.

Since this issue will be closed once `additionalRefHooks` lands, feel free to open a separate issue for the mutable external store case so we can keep tracking it.

For reference, the full schema for custom hook configs in React Compiler ([`Environment.ts`](https://github.com/facebook/react/blob/81e442eaf3fbfabd2367a0d8691ec0df13a3add0/compiler/packages/babel-plugin-react-compiler/src/HIR/Environment.ts#L94)):

```ts
const HookSchema = z.object({
  /*
   * The effect of arguments to this hook. Describes whether the hook may or may
   * not mutate arguments, etc.
   */
  effectKind: z.nativeEnum(Effect),

  /*
   * The kind of value returned by the hook. Allows indicating that a hook returns
   * a primitive or already-frozen value, which can allow more precise memoization
   * of callers.
   */
  valueKind: z.nativeEnum(ValueKind), // "maybefrozen" | "frozen" | "primitive" | "global" | "mutable" | "context"

  /*
   * Specifies whether hook arguments may be aliased by other arguments or by the
   * return value of the function. Defaults to false. When enabled, this allows the
   * compiler to avoid memoizing arguments.
   */
  noAlias: z.boolean().default(false),

  /*
   * Specifies whether the hook returns data that is composed of:
   * - undefined
   * - null
   * - boolean
   * - number
   * - string
   * - arrays whose items are also transitiveMixed
   * - objects whose values are also transitiveMixed
   *
   * Many state management and data-fetching APIs return data that meets
   * this criteria since this is JSON + undefined. Forget can compile
   * hooks that return transitively mixed data more optimally because it
   * can make inferences about some method calls (especially array methods
   * like `data.items.map(...)` since these builtin types have few built-in
   * methods.
   */
  transitiveMixedData: z.boolean().default(false),
});
```

**Comment by silverwind** [comment]
Not sure if type-aware linting is possible in those codebase buf if a hook returns `React.Ref` or `React.RefObject`, that's a definitive indicator that a variable is a ref. If that works, I wouldn't need `additionalRefHooks`.

**Comment by Rel1cx** [comment]
> Not sure if type-aware linting is possible in those codebase buf if a hook returns `React.Ref` or `React.RefObject`, that's a definitive indicator that a variable is a ref. If that works, I wouldn't need `additionalRefHooks`.

Yes, it's possible. The current `react-x/no-implicit-ref` rule already performs similar checks using type information.

I also considered converting `react-x/immutability` to a type-aware check when thinking about solutions for the `useWorld` example above. With that approach, external mutable store hooks would not need an `additionalMutableHooks` setting either, because it is straightforward to distinguish via the fully qualified signature whether a method belongs to (or inherits from) native types like `Array`, `Map`, or `Set`, or is a custom method with the same name on an external mutable store, such as [the `set` method on the store returned by `useWorld()` in `koota`](https://github.com/pmndrs/koota/blob/4468d1e5ebf44db4c6d5303f56c321d3aef07846/packages/core/src/world/types.ts#L62).

However, `additionalRefHooks` would benefit at least 3 to 4 existing rules that do not rely on type checking, in addition to `react-x/immutability`. Introducing type checking into all of those rules to keep things consistent would be expensive and a breaking change, so the `additionalRefHooks` option still has value.

Furthermore, since [it's still unclear how well `typescript-eslint` will perform with type checking in TS 7](https://github.com/typescript-eslint/typescript-eslint/issues/10940#issuecomment-4922812181), I prefer to avoid converting more existing rules to type-aware at the moment.
